src/darwin/Framework/CHIP/CHIPPersistentStorageDelegateBridge.h

```diff
@@ -30,21 +30,16 @@ class CHIPPersistentStorageDelegateBridge : public chip::PersistentStorageDelega
 
     void setFrameworkDelegate(_Nullable id<CHIPPersistentStorageDelegate> delegate, _Nullable dispatch_queue_t queue);
 
-    void SetStorageDelegate(chip::PersistentStorageResultDelegate * delegate) override;
```

Comment:
Given this is Darwin, we should keep this all async, sync doesn't match the style of the platform

Reply:
At what level should I do the split?

The interface that chip will use is sync calls. I was told that KVS is sync because other underlying platforms (zephyr and nvs) are synchronous and I followed that pattern. I figured in any case switching between sync/async should be simple enough once the interface is stable. 

Please suggest an approach here - my darwin development-fu is not that great.

Reply:
> > [#6272 (review)](https://github.com/project-chip/connectedhomeip/pull/6272#pullrequestreview-646497163)
> 
> Replied there and also followed up offline. Generally:
> 
> * Agree that async is more flexible. Thinking we can create an IO thread for whenever we need async behaviour
> * current goal is to help untangle the persistent storage story (have configurationamanger, kvs, storage delegate, then have sync and async in various implementation state, have async callbacks not used)
> 
> This PR is starting to consolidate the APIs up to a point of at least having a uniform API surface for storage
> 
> Future PRs expected with:
> - consolidate into one storage class (I expect KVS and replace storage delegate and configuration manager)
> - move KVS outside of platform (Android does not like that) and provide pluggability
> - add IO thread for async storage needs

What do you mean here? Create a new thread and synchronize on it with a blocking primitive like a conditional variable?

I think this is backwards and would suggest that the synchronous APIs be removed from the core in favor of the asynchronous ones. Hooking the async high level API up to a synchronous platform backed can be handled using a background thread. With this approach the core protocol need not block; i.e. we're not forcing storage operations to be "stop the world" events.

Note we already have an IO thread; that's where we incoming packets are handled.

> - start taking advantage of ability of binary storage - do not double-base64 anymore and look into some encoding/decoding uniformity (have at least header, versioning and endianess flags) for binary data

Reply:
I can go the 'make everything async' path as well. Right now I have:

- KVS has consistent API on all platforms and uses void * (I like this) but is sync (not the best)
- PersistentStorageDelegate has inconsistent API surface and uses char * only (do not like this) but some implementations have async version (more flexible)


I am currently only concerned about uniform API access and having binary data formatting rather than attempts at base64 and sprintf number formatting because of char*. I can followup with async either way.

Making KVS async is a somewhat larger task.

Regarding the IO thread,I believe we are callback based right now for IO.

Reply:
@woody-apple - this is implementing a CHIP interface under darwin. The CHIP interface asks for sync, so it is not a platform choice regarding sync/async.

Are you suggesging an additional separator? The 'GetKeyValue' was already sync before (and this is my main complaint with the current interface: it is not consistently using sync vs async).